docs/okeps/okep-5377-extend-udn-to-support-multiple-cluster-subnets-in-layer3-topology.md

```diff
@@ -0,0 +1,284 @@
+   interface. For example, create UDN with one cluster subnet:
+
+   When a new subnet is added to a secondary Layer3 network, existing Pods do not receive updated
+   routes to reach the new subnet. As a result, traffic to IPs in the new subnet is routed via the
+   primary network interface. The issue is depicted as below:
+
+   - A Secondary Layer3 network is created with one cluster subnet:
+      ```yaml
+      apiVersion: k8s.ovn.org/v1
+      kind: UserDefinedNetwork
+      metadata:
+         name: udn-secondary-layer3
+      spec:
+         topology: Layer3
+         layer3:
+            role: Secondary
+            subnets:
+            - cidr: 10.10.0.0/16
+            hostSubnet: 17
+      ```
+   - Node A allocates a subnet from `10.10.0.0/16`:
+      ```bash
+         k8s.ovn.org/node-subnets: '{...,"udn_udn-secondary-layer3":["10.10.128.0/17"]}'
+      ```
+   - Pod A is created on Node A and gets the following secondary network config:
+      ```
+      k8s.ovn.org/pod-networks: '{
+         ...,
+         "udn/udn-secondary-layer3":{
+            "ip_addresses":["10.10.128.4/17"],"mac_address":"0a:58:0a:0a:80:04",
+            "routes":[
+                  {"dest":"10.10.0.0/16","nextHop":"10.10.128.1"}],
+            "ip_address":"10.10.128.4/17","role":"secondary"}}'
+      ```
+   - A new subnet is added to the UDN:
+      ```
+      apiVersion: k8s.ovn.org/v1
+      kind: UserDefinedNetwork
+      metadata:
+      name: udn-secondary-layer3
+      namespace: udn
+      spec:
+      layer3:
+         role: Secondary
+         subnets:
+         - cidr: 10.10.0.0/16
+            hostSubnet: 17
+         - cidr: 10.20.0.0/16
+            hostSubnet: 17
+      topology: Layer3
+      ```
+   - Node B allocates a subnet from `10.20.0.0/16`:
+      ```
+      k8s.ovn.org/node-subnets: '{"default":["10.192.2.0/24"],"udn_udn-primary-layer3":["10.2.0.0/17"],"udn_udn-secondary-layer3":["10.20.0.0/17"]}'
+      ```
+   - Pod B is created on Node B and receives both route entries:
+      ```
+      k8s.ovn.org/pod-networks: '{
+         ...,
+         "udn/udn-secondary-layer3":{
+            "ip_addresses":["10.20.0.4/17"],"mac_address":"0a:58:0a:14:00:04",
+            "routes":[
+                  {"dest":"10.10.0.0/16","nextHop":"10.20.0.1"},
+                  {"dest":"10.20.0.0/16","nextHop":"10.20.0.1"}],
+            "ip_address":"10.20.0.4/17","role":"secondary"}}'
+      ```
+   - traffic from Pod A to Pod B is routed via the primary network:
+      ```bash
+      root@client-a:/# ip route get 10.20.0.4
+      10.20.0.4 via 10.1.0.1 dev ovn-udn1 src 10.1.0.10 uid 0
+      cache
+      ```
+
+   This issue occurs because the existing Pod (Pod A) lacks route updates to the newly added
+   subnet. It can be resolved by recreating affected Pods.
+
+
+   **Note**: This issue does not affect the primary network, as its interface is always the default
```

Comment:
Should this feature be scoped to primary networks only? recreating all pods is almost the same as recreating the network itself, so that doesn't seem beneficial?

Reply:
I have the same question

Reply:
Make sense, I will scope it to primary network only.